From 4685ce879aaf908cbacee2196eab1b33b8bbf698 Mon Sep 17 00:00:00 2001
From: Jason Liu <jason.hui.liu@nxp.com>
Date: Thu, 31 Oct 2019 14:48:12 +0800
Subject: [PATCH 1/2] MLK-16005-2 soc: imx: scu: add the SW workaround for
 i.MX8QM TKT340553

on i.MX8QM 1.0/1.1,TLB maintenance through DVM messages over ARADDR channel,
some bits (see the following) will be corrupted:

ASID[15:12] VA[48:45] VA[44:41] VA[39:36]

This issue will result in the TLB aintenance across the clusters not working
as expected due to some VA and ASID bits get corrupted

The SW workaround is: use the vmalle1is if VA larger than 36bits or
ASID[15:12] is not zero, otherwise, we use original TLB maintenance path.

Note: To simplify the code, we did not check VA[40] bit specifically

[ Leo: Separated arch changes into a separate patch ]

Upstream-Status: Pending
Signed-off-by: Jason Liu <jason.hui.liu@nxp.com>
Reviewed-by: Anson Huang <anson.huang@nxp.com>
---
 drivers/firmware/imx/imx-scu-soc.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/drivers/firmware/imx/imx-scu-soc.c b/drivers/firmware/imx/imx-scu-soc.c
index 497192320562..cea7576ec175 100644
--- a/drivers/firmware/imx/imx-scu-soc.c
+++ b/drivers/firmware/imx/imx-scu-soc.c
@@ -12,6 +12,8 @@
 
 static struct imx_sc_ipc *imx_sc_soc_ipc_handle;
 
+extern bool TKT340553_SW_WORKAROUND;
+
 struct imx_sc_msg_misc_get_soc_id {
 	struct imx_sc_rpc_msg hdr;
 	union {
@@ -82,6 +84,7 @@ static const char *imx_scu_soc_name(u32 id)
 {
 	switch (id) {
 	case 0x1:
+		TKT340553_SW_WORKAROUND = true;
 		return "i.MX8QM";
 	case 0x2:
 		return "i.MX8QXP";
-- 
2.47.3

